File: src/cloudflare/transport.rs
use std::rc::Rc;
use futures::lock::Mutex;
use worker::*;
use crate::cloudflare::connection::CloudflareConnectionTransport;
use crate::cloudflare::create_cors_response;
use crate::cloudflare::id_allocator::{DoIdBatchStore, IdAllocator};
use crate::cloudflare::kv::CloudflareKvStore;
use crate::cloudflare::rpc_context::CloudflareRpcContext;
use crate::cloudflare::storage::CloudflareStorage;
use crate::nostr::connection::ConnectionManager;
use crate::nostr::engine::NostrEngine;
use crate::util::short;
use crate::log_info;
use crate::log_debug;
use crate::log_warn;
use crate::log_error;

#[durable_object]
pub struct CloudflareTransport {
    env: Env,
    engine: Rc<Mutex<NostrEngine<CloudflareStorage>>>,
    manager: Rc<Mutex<ConnectionManager<CloudflareConnectionTransport>>>,
    id_allocator: Rc<IdAllocator<DoIdBatchStore>>,
    kv: CloudflareKvStore,
    rpc_ctx: CloudflareRpcContext,
}

impl DurableObject for CloudflareTransport {
    fn new(state: State, env: Env) -> Self {
        let state = Rc::new(state);
        let storage = CloudflareStorage::new(state.storage());
        let max_content_length = env.var("MAX_CONTENT_LENGTH")
            .and_then(|v| v.to_string().parse::<usize>().map_err(|e| Error::from(e.to_string())))
            .unwrap_or(65536);
        let engine = NostrEngine::new_with_storage(storage, max_content_length, crate::util::now);
        let kv = CloudflareKvStore::new(env.kv("MEKHALA_NWC_KV").expect("MEKHALA_NWC_KV not configured"));

        let transport = CloudflareConnectionTransport::new(Rc::clone(&state));
        let manager = ConnectionManager::new(transport);
        let id_store = DoIdBatchStore::new(state.storage());
        let id_allocator = IdAllocator::new(id_store);

        let engine = Rc::new(Mutex::new(engine));
        let manager = Rc::new(Mutex::new(manager));
        let id_allocator = Rc::new(id_allocator);
        let rpc_ctx = CloudflareRpcContext::new(
            Rc::clone(&engine),
            Rc::clone(&manager),
            Rc::clone(&id_allocator),
        );

        Self {
            env,
            engine,
            manager,
            id_allocator,
            kv,
            rpc_ctx,
        }
    }

    async fn fetch(&self, req: Request) -> Result<Response> {
        let url = req.url()?;
        let path = url.path();

        if path.starts_with("/lnaddress/") && path.ends_with("/callback") {
            let username = path.strip_prefix("/lnaddress/").and_then(|s| s.strip_suffix("/callback")).unwrap_or("");
            let handler = crate::lnaddress::LnAddressHandler::new(&self.kv);
            return handler.handle_callback(req, username, self).await;
        }

        self.accept_new_connection().await
    }

    async fn websocket_message(&self, websocket: WebSocket, message: WebSocketIncomingMessage) -> Result<()> {
        log_debug!("→ websocket_message handler invoked");
        let text = match message {
            WebSocketIncomingMessage::String(t) => t,
            _ => {
                log_warn!("binary message not supported");
                let _ = websocket.send_with_str(&crate::nostr::RelayMessage::Notice("binary not supported".into()).to_json());
                return Ok(());
            }
        };

        if text.len() > 131072 {
            log_warn!("message too large: {} bytes", text.len());
            let _ = websocket.send_with_str(&crate::nostr::RelayMessage::Notice("message too large".into()).to_json());
            return Ok(());
        }

        let connection_id = match self.manager.lock().await.identify(&websocket) {
            Some(id) => id,
            None => {
                log_error!("connection not found, sending reconnect notice");
                let _ = websocket.send_with_str(&crate::nostr::RelayMessage::Notice("connection lost: please reconnect".into()).to_json());
                return Ok(());
            }
        };

        let mut engine = self.engine.lock().await;
        let engine_ref = &mut *engine;

        self.manager.lock().await.wake_and_load(connection_id, engine_ref).await;

        let parsed = crate::nostr::ClientMessage::from_json(&text);
        match &parsed {
            Ok(crate::nostr::ClientMessage::Event(event)) => {
                log_info!("← conn={} RECV EVENT kind={} pk={} id={}", connection_id, event.kind, short(&event.pubkey, 8), short(&event.id, 8));
            }
            Ok(crate::nostr::ClientMessage::Req(sub_id, filters)) => {
                log_info!("← conn={} RECV REQ sub={}", connection_id, sub_id);
                for (i, f) in filters.iter().enumerate() {
                    let kinds = f.kinds.as_ref().map(|k| format!("{:?}", k)).unwrap_or_else(|| "any".into());
                    let auths = f.authors.as_ref().map(|a| a.iter().map(|s| short(s, 8)).collect::<Vec<_>>().join(",")).unwrap_or_else(|| "".into());
                    let pts = f.p_tags.as_ref().map(|p| p.iter().map(|s| short(s, 8)).collect::<Vec<_>>().join(",")).unwrap_or_else(|| "".into());
                    log_info!("  filter[{}]: kinds={} authors={} #p={}", i, kinds, auths, pts);
                }
            }
            Ok(crate::nostr::ClientMessage::Close(sub_id)) => {
                log_info!("← conn={} RECV CLOSE sub={}", connection_id, sub_id);
            }
            Err(parse_error) => {
                log_error!("✗ parse failed: {}", parse_error.message);
            }
        }

        let responses = match parsed {
            Ok(msg) => engine_ref.handle_typed(connection_id, msg).await,
            Err(parse_error) => parse_error.into_responses(connection_id),
        };

        self.manager.lock().await.dispatch(responses, engine_ref).await;
        Ok(())
    }

    async fn websocket_close(&self, ws: WebSocket, code: usize, reason: String, was_clean: bool) -> Result<()> {
        log_debug!("→ websocket_close handler invoked code={} reason={} clean={}", code, reason, was_clean);
        self.handle_disconnect(&ws).await
    }

    async fn websocket_error(&self, ws: WebSocket, error: Error) -> Result<()> {
        log_debug!("→ websocket_error handler invoked err={}", error);
        self.handle_disconnect(&ws).await
    }
}

#[async_trait::async_trait(?Send)]
impl crate::common::NwcTransport for CloudflareTransport {
    async fn get_wallet_info(&self, pubkey: &str) -> Option<crate::nostr::WalletInfo> {
        let mut engine = self.engine.lock().await;
        let conn_ids = engine.load_by_pubkey(pubkey).await;

        {
            let manager = self.manager.lock().await;
            for id in conn_ids {
                manager.try_activate(id);
            }
        }

        engine.get_wallet_info(pubkey).await
    }

    async fn execute_nwc_rpc(&self, request: crate::nostr::Event) -> Result<crate::nostr::Event, crate::common::NwcError> {
        crate::nostr::rpc_orchestrator::execute_nwc_rpc(&self.rpc_ctx, request).await
    }
}

impl CloudflareTransport {
    async fn accept_new_connection(&self) -> Result<Response> {
        let max_connections = self.env.var("MAX_CONNECTIONS")
            .and_then(|v| v.to_string().parse::<usize>().map_err(|e| Error::from(e.to_string())))
            .unwrap_or(20);

        let WebSocketPair { client, server } = WebSocketPair::new()?;

        // 1. Allocate ID without holding any borrow across await
        let connection_id = self.id_allocator.allocate().await;

        // 2. Accept + register atomically with count re-check inside lock
        {
            let mut manager = self.manager.lock().await;
            if manager.total_count() >= max_connections {
                log_warn!("✗ connection rejected: max={} connections reached", max_connections);
                return Response::error("Too Many Requests", 429);
            }
            manager.accept_and_register(connection_id, &server);
        }

        // 3. NOW safe to await — WS is accepted and in HashMap
        let mut engine = self.engine.lock().await;
        let engine_ref = &mut *engine;
        let responses = engine_ref.on_connect(connection_id).await;

        let response = Response::from_websocket(client)?;

        let mut manager = self.manager.lock().await;
        log_info!("+ conn={} accepted count={}/{}", connection_id, manager.total_count(), max_connections);
        manager.dispatch(responses, engine_ref).await;

        Ok(response)
    }

    async fn handle_disconnect(&self, ws: &WebSocket) -> Result<()> {
        let conn_id = self.manager.lock().await.identify(ws);

        if let Some(id) = conn_id {
            let mut engine = self.engine.lock().await;
            let engine_ref = &mut *engine;
            let mut manager = self.manager.lock().await;

            // Wake up if hibernated, then terminate
            manager.wake_and_load(id, engine_ref).await;

            log_info!("- conn={} disconnected", id);
            manager.on_terminate(id, engine_ref).await;
        }
        Ok(())
    }
}

pub async fn connect(req: Request, env: &Env) -> Result<Response> {
    crate::cloudflare::get_durable_stub(env)?.fetch_with_request(req).await
}

pub fn create_response(info: serde_json::Value, content_type: &str) -> Result<Response> {
    let mut response = create_cors_response(Response::from_json(&info)?)?;
    response.headers_mut().set("Content-Type", content_type)?;
    Ok(response)
}

File: src/cloudflare/transport.ts
import { tagsArrayFromJSON, type JsonValue } from '../nostr/tag.ts';
import type { Event } from '../nostr/event.ts';
import type { NwcUri } from '../nostr/nip47.ts';
import type { ClientMessage, RelayMessage } from '../nostr/nip01.ts';
import { parseClientMessage, relayMessageToJSON } from '../nostr/nip01.ts';
import { NostrEngine } from '../nostr/engine.ts';
import type { EngineResponse } from '../nostr/engine.ts';
import { CloudflareStorage } from './storage.ts';
import { fromEnv } from './config.ts';
import type { CloudflareConfig } from './config.ts';
import { ConnectionRegistry } from './connection.ts';
import type { WebSocketHandle } from './connection.ts';
import { CloudflareKvStore } from './kv.ts';
import { NwcRpcMachine } from '../nostr/rpc_machine.ts';
import type { RpcAction } from '../nostr/rpc_machine.ts';
import { NwcError } from '../common/mod.ts';
import type { WalletInfo } from '../nostr/nip47.ts';
import { NwcClient, parseNwcUri, EncryptionMethod } from '../nostr/nip47.ts';
import { DEFAULT_LIMITS } from '../nostr/limits.ts';

const RPC_TIMEOUT_MS = 10_000;
const MAX_MESSAGE_LENGTH = 131_072;

function now(): number {
  return Math.floor(Date.now() / 1000);
}

export class CloudflareTransport implements DurableObject {
  private config: CloudflareConfig;
  private engine: NostrEngine<CloudflareStorage>;
  private connections = new ConnectionRegistry();
  private idCounter = 0;
  private kv: CloudflareKvStore;
  private ctx: DurableObjectState;
  // Pending LNURL callback: { resolve, requestId, deadline }
  private pendingCallback: { resolve: (val: unknown) => void; requestId: string; deadline: number } | null = null;

  constructor(ctx: DurableObjectState, env: Record<string, unknown>) {
    this.ctx = ctx;
    this.config = fromEnv(env as unknown as Record<string, string | undefined>);
    const storage = new CloudflareStorage(ctx.storage);
    const limits = {
      ...DEFAULT_LIMITS,
      maxContentLength: this.config.maxContentLength,
      maxSubscriptionsPerConnection: this.config.maxSubscriptionsPerConnection,
    };
    this.engine = new NostrEngine(storage, limits, now);
    this.kv = new CloudflareKvStore(
      (env as Record<string, unknown>).MEKHALA_NWC_KV as KVNamespace,
    );
  }

  // ── DurableObject lifecycle ──

  async fetch(request: Request): Promise<Response> {
    const url = new URL(request.url);
    const path = url.pathname;

    if (path.includes('/lnaddress/') && path.endsWith('/callback')) {
      return this.handleLnaddressCallback(request, path);
    }

    return this.acceptNewConnection();
  }

  async webSocketMessage(ws: WebSocket, message: ArrayBuffer | string): Promise<void> {
    if (typeof message !== 'string') {
      ws.send(JSON.stringify(['NOTICE', 'binary not supported']));
      return;
    }

    if (message.length > MAX_MESSAGE_LENGTH) {
      ws.send(JSON.stringify(['NOTICE', 'message too large']));
      return;
    }

    // Before processing, try to detect NWC responses for pending callbacks
    if (message.includes('"kind":23195') || message.includes('"kind":23196') || message.includes('"kind":23197')) {
      try {
        const parsed = JSON.parse(message);
        if (Array.isArray(parsed) && parsed[0] === 'EVENT') {
          const rawEvent = parsed[1] as { tags?: string[][]; id?: string };
          if (rawEvent?.tags) {
            // Check pending callback first (non-blocking, no storage I/O)
            for (const tag of rawEvent.tags) {
              if (tag[0] === 'e' && tag[1]) {
                const eid = tag[1];
                if (this.pendingCallback && this.pendingCallback.requestId === eid) {
                  this.pendingCallback.resolve(rawEvent);
                  this.pendingCallback = null;
                  break;
                }
              }
            }
            // Also store in DO storage as fallback
            for (const tag of rawEvent.tags) {
              if (tag[0] === 'e' && tag[1]) {
                const eid = tag[1];
                const rpcKey = `lnrpc:${eid}`;
                const pending = await this.ctx.storage.get(rpcKey);
                if (pending) {
                  await this.ctx.storage.put(`lnrpc_result:${eid}`, parsed[1]);
                  break;
                }
              }
            }
          }
        }
      } catch { /* ignore parse errors */ }
    }

    await processMessage(message, ws, this.engine, this.connections);
  }

  async webSocketClose(ws: WebSocket, _code: number, _reason: string, _wasClean: boolean): Promise<void> {
    await this.handleDisconnect(ws);
  }

  async webSocketError(ws: WebSocket, _error: unknown): Promise<void> {
    await this.handleDisconnect(ws);
  }

  // ── NwcTransport methods ──

  async getWalletInfo(pubkey: string): Promise<WalletInfo | null> {
    await this.loadConnectionByPubkey(pubkey);
    return this.engine.getWalletInfo(pubkey);
  }

  async executeNwcRpc(request: Event): Promise<Event> {
    const id = this.allocateId();
    const machine = new NwcRpcMachine(request);

    for (const action of machine.start()) {
      await this.executeRpcActionInner(id, action);
    }

    const result = this.receiveRpcResponse(id, machine);

    await this.engine.onDisconnect(id);
    this.connections.remove(id);

    return result;
  }

  // ── ID allocation ──

  private allocateId(): number {
    this.idCounter++;
    return this.idCounter;
  }

  // ── Internal RPC ──

  private async receiveRpcResponse(id: number, machine: NwcRpcMachine): Promise<Event> {
    const start = Date.now();

    while (true) {
      const elapsed = Date.now() - start;
      const remaining = RPC_TIMEOUT_MS - elapsed;
      if (remaining <= 0) {
        throw NwcError.timeout();
      }

      let text: string;
      try {
        text = await Promise.race([
          this.connections.addInternal(id),
          new Promise<string>((_, reject) =>
            setTimeout(() => reject(NwcError.timeout()), remaining),
          ),
        ]);
      } catch (e) {
        if (e instanceof NwcError) throw e;
        throw NwcError.timeout();
      }

      let msg: RelayMessage;
      try {
        msg = JSON.parse(text) as RelayMessage;
      } catch {
        throw NwcError.protocolError('malformed relay response');
      }

      const action = machine.transition(msg);
      if (action) {
        await this.executeRpcActionInner(id, action);
      }

      const state = machine.getState();
      if (state.kind === 'success') {
        return state.event;
      }
      if (state.kind === 'failed') {
        throw NwcError.protocolError(state.reason);
      }
    }
  }

  private async executeRpcActionInner(id: number, action: RpcAction): Promise<void> {
    switch (action.kind) {
      case 'subscribe': {
        const responses = await this.engine.handleReqInternal(id, action.subId, [action.filter]);
        if (responses.some(r => r.kind === 'send' && r.message.type === 'CLOSED')) {
          throw NwcError.protocolError('subscribe rejected: storage unavailable');
        }
        this.sendToWsOnly(responses);
        break;
      }
      case 'publish': {
        const msg: ClientMessage = { type: 'EVENT', event: action.event };
        const responses = await this.engine.handleTyped(id, msg);
        this.sendToWsOnly(responses);
        break;
      }
      case 'unsubscribe': {
        const responses = await this.engine.processClose(id, action.subId);
        this.sendToWsOnly(responses);
        break;
      }
    }
  }

  // ── Response dispatch ──

  private recoverConnections(): void {
    const websockets = this.ctx.getWebSockets();
    for (const ws of websockets) {
      this.connections.identify(ws);
    }
  }

  private sendToWsOnly(responses: EngineResponse[]): void {
    for (const resp of responses) {
      if (resp.kind === 'send') {
        this.connections.send(resp.recipientId, relayMessageToJSON(resp.message));
      }
    }
  }

  // ── Connection management ──

  private async loadConnectionByPubkey(pubkey: string): Promise<void> {
    const ids = await this.engine.loadByPubkey(pubkey);
    if (ids.length > 0) {
      await this.engine.load(ids[0]);
    }
  }

  private async acceptNewConnection(): Promise<Response> {
    if (this.connections.size >= this.config.maxConnections) {
      return new Response(JSON.stringify({ status: 'ERROR', reason: 'Too Many Connections' }), {
        status: 429,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const pair = new WebSocketPair();
    const client = pair[0];
    const server = pair[1];

    const connectionId = this.allocateId();

    this.ctx.acceptWebSocket(server);
    this.connections.addExternal(connectionId, server);

    const responses = await this.engine.onConnect(connectionId);
    this.sendToWsOnly(responses);

    return new Response(null, { status: 101, webSocket: client });
  }

  private async handleDisconnect(ws: WebSocket): Promise<void> {
    const id = this.connections.identify(ws);
    if (id == null) return;

    const responses = await this.engine.onTerminate(id);
    this.sendToWsOnly(responses);
    this.connections.remove(id);
  }

  private async handleLnaddressCallback(request: Request, path: string): Promise<Response> {
    const username = path.substring(path.indexOf('/lnaddress/') + '/lnaddress/'.length, path.lastIndexOf('/callback'));
    const url = new URL(request.url);
    const amountMsat = parseInt(url.searchParams.get('amount') || '', 10);
    if (isNaN(amountMsat) || amountMsat <= 0) {
      return jsonResponse({ status: 'ERROR', reason: 'Missing amount' }, 200);
    }

    const nwcUriStr = await this.kv.getNwcUri(username);
    if (nwcUriStr == null) {
      return jsonResponse({ status: 'ERROR', reason: 'User not found' }, 200);
    }

    let nwcUri: NwcUri;
    try {
      nwcUri = parseNwcUri(nwcUriStr);
    } catch {
      return jsonResponse({ status: 'ERROR', reason: 'Invalid NWC URI' }, 200);
    }

    const client = new NwcClient(nwcUri);

    // Detect wallet's preferred encryption method from its cached info event
    const walletInfo = await this.engine.getWalletInfo(nwcUri.walletPubkey);
    if (walletInfo?.encryptionAlgorithms.includes(EncryptionMethod.Nip44)) {
      client.encryptionMethod = EncryptionMethod.Nip44;
    }

    const { event } = await client.createRequestEvent('make_invoice', { amount: amountMsat }, []);

    // Recover WebSocket connections — DO may have hibernated since last handler
    this.recoverConnections();

    const rpcId = this.allocateId();
    const subId = 'lnrpc';

    const subResponses = await this.engine.handleReqInternal(rpcId, subId, [
      { eTags: [event.id], pTags: [event.pubkey] },
    ]);
    this.sendToWsOnly(subResponses);

    const pubResponses = await this.engine.handleTyped(rpcId, { type: 'EVENT', event });
    this.sendToWsOnly(pubResponses);

    const rpcKey = `lnrpc:${event.id}`;
    await this.ctx.storage.put(rpcKey, { walletPubkey: nwcUri.walletPubkey, clientPubkey: event.pubkey, createdAt: Date.now() });

    // Wait for the wallet's response via a Promise resolved by webSocketMessage.
    // This avoids busy-polling and lets the DO runtime process WebSocket messages.
    const rawEvent = await new Promise<unknown>((resolve) => {
      this.pendingCallback = { resolve, requestId: event.id, deadline: Date.now() + RPC_TIMEOUT_MS };
      // Fallback: resolve via timeout
      setTimeout(() => {
        if (this.pendingCallback?.requestId === event.id) {
          this.pendingCallback = null;
          resolve(null);
        }
      }, RPC_TIMEOUT_MS);
    });

    if (rawEvent) {
      await this.ctx.storage.delete(rpcKey);
      await this.ctx.storage.delete(`lnrpc_result:${event.id}`);
      await this.engine.processClose(rpcId, subId);
      try {
        const raw = rawEvent as Record<string, unknown>;
        const parsedEvent: Event = {
          id: raw.id as string,
          pubkey: raw.pubkey as string,
          createdAt: (raw.created_at ?? raw.createdAt) as number,
          kind: raw.kind as number,
          tags: tagsArrayFromJSON((raw.tags as JsonValue[][] | undefined) ?? []),
          content: raw.content as string,
          sig: raw.sig as string,
        };
        // Manually decrypt and parse the response to avoid verifyEvent issues
        const decryptedContent = await client.decrypt(parsedEvent.content);
        const responseJson = JSON.parse(decryptedContent) as { result?: Record<string, unknown> };
        const invoice = responseJson?.result?.invoice;
        if (invoice) {
          return jsonResponse({ pr: invoice, routes: [] }, 200);
        }
      } catch {
        // fall through to error
      }
      return jsonResponse({ status: 'ERROR', reason: 'Failed to parse wallet response' }, 200);
    }

    await this.ctx.storage.delete(rpcKey);
    await this.engine.processClose(rpcId, subId);
    return jsonResponse({ status: 'ERROR', reason: 'Wallet not connected' }, 200);
  }
}

// ── Helpers ──

function jsonResponse(body: unknown, status: number): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, OPTIONS',
      'Access-Control-Allow-Headers': '*',
      'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
      'X-Content-Type-Options': 'nosniff',
    },
  });
}

// ── Response dispatch helper ──

function sendToConnection(
  connections: ConnectionRegistry,
  fallbackWs: WebSocketHandle,
  responses: EngineResponse[],
): void {
  for (const resp of responses) {
    if (resp.kind === 'send') {
      const sent = connections.send(resp.recipientId, relayMessageToJSON(resp.message));
      if (!sent) {
        fallbackWs.send(relayMessageToJSON(resp.message));
      }
    }
  }
}

// ── Extracted message processing ──

export async function processMessage(
  text: string,
  ws: WebSocketHandle,
  engine: NostrEngine<CloudflareStorage>,
  connections: ConnectionRegistry,
): Promise<void> {
  let parsed: ClientMessage;
  try {
    parsed = parseClientMessage(text);
  } catch (e) {
    const errMsg = e instanceof Error ? e.message : String(e);
    try {
      const partial = JSON.parse(text);
      if (Array.isArray(partial) && partial[0] === 'EVENT' && partial[1]?.id) {
        ws.send(JSON.stringify(['OK', partial[1].id, false, `parse failed: ${errMsg}`]));
      } else {
        ws.send(JSON.stringify(['NOTICE', `parse failed: ${errMsg}`]));
      }
    } catch {
      ws.send(JSON.stringify(['NOTICE', `parse failed: ${errMsg}`]));
    }
    return;
  }

  const connectionId = connections.identify(ws);
  if (connectionId == null) {
    ws.send(JSON.stringify(['NOTICE', 'connection lost: please reconnect']));
    return;
  }
  await engine.load(connectionId);
  connections.addExternal(connectionId, ws);

    if (parsed.type === 'EVENT') {
      const event = parsed.event;
      const result = engine.validateEvent(event);
      if (result.ok) {
        ws.send(relayMessageToJSON({ type: 'OK', id: event.id, ok: true, message: '' }));
        const responses = await engine.routeVerifiedEvent(connectionId, event);
        sendToConnection(connections, ws, responses);
      } else {
        ws.send(relayMessageToJSON({ type: 'OK', id: result.id, ok: false, message: result.error }));
      }
    } else {
      const responses = await handleNonEventMessage(parsed, connectionId, engine);
      sendToConnection(connections, ws, responses);
    }
}

async function handleNonEventMessage(
  msg: ClientMessage,
  connectionId: number,
  engine: NostrEngine<CloudflareStorage>,
): Promise<EngineResponse[]> {
  switch (msg.type) {
    case 'REQ':
      return engine.handleReq(connectionId, msg.subscriptionId, msg.filters);
    case 'CLOSE':
      return engine.processClose(connectionId, msg.subscriptionId);
    default:
      return [];
  }
}
